LLama/Experimental/Abstractions/ITokenizer.cs

```diff
@@ -0,0 +1,34 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Text;
+
+namespace LLama.Experimental.Abstractions
+{
+    /// <summary>
+    /// The interface for tokenizer in LLamaSharp. It's responsible for converting text to token ids, or vice versa.
+    /// </summary>
+    public interface ITokenizer
+    {
+        // TODO: `ApplyChatTemplate` API
+
+        // TODO: Batched Encode?
+
+        /// <summary>
+        /// Get the token ids from the text
+        /// </summary>
+        /// <param name="input"></param>
+        /// <returns></returns>
+        IList<int> Tokenize(string input);
+
+        /// <summary>
+        /// Convert the token ids to text.
+        /// </summary>
+        /// <param name="tokenIds"></param>
+        /// <param name="result"></param>
+        /// <param name="skipSpecialTokens"></param>
+        /// <returns>The consumed tokens for decoding.</returns>
+        int ConvertIdsToText(IEnumerable<int> tokenIds, out string result, bool skipSpecialTokens = false);
```

Comment:
We can't have an API like this. It's the same problem as the old `DeTokenize` method, sovled by the `StreamingTokenDecoder`.

A single character may be several tokens. So for example say the tokens `[1, 2, 3]` produce the character `A` then decoding `[1, 2]` will produce a broken string. Then decoding `[3]` will produce another broken string.

Reply:
There is an `IncrementalDecodingOffset` member to [Sequence](https://github.com/SciSharp/LLamaSharp/pull/683/files#diff-b356b13cabe67f98effd3141ba05e471ad296df3afe3e2794e44c4d702874a6fR64). I'm trying to find a way to integrate the streaming decoder better into this design but haven't made it yet. The biggest problem of using it directly is that it rents some memories and needs to be released. Though adding `StreamingTokenDecoder` as a member of `Sequence` could make it auto-released, it's a bit weird to make `Sequence` own a `StreamingTokenDecoder` because it should be a part of the tokenizer. Besides, the `Sequence`, as a mid-level class which may used frequently by users, is not supposed to deal with logics related with `LLamaContext`, while  `LLamaContext` is required to initialize a `StreamingTokenDecoder`.